### PR TITLE
Bug 1355169 - Add rate-limiting to show_bug.cgi and rest.cgi

### PR DESCRIPTION
**Pull request by umohm1**

### CODE REVIEW

File: Bugzilla.pm [hunk 1]
```diff
@@ -155,7 +156,7 @@ sub init_page {
     }
 
     # If Bugzilla is shut down, do not allow anything to run, just display a
-    # message to the user about the downtime and log out.  Scripts listed in 
+    # message to the user about the downtime and log out.  Scripts listed in
```
Comment (floatingatoll) [inline Bugzilla.pm]: If at all possible, please extract the whitespace cleanup into its own commit. It's okay if it's in the pull request, but it shouldn't be mixed with a functionality-altering commit unless that specific code is being modified.

File: Bugzilla.pm [hunk 2]
```diff
@@ -54,6 +54,7 @@ use File::Basename;
 use File::Spec::Functions;
 use Safe;
 use Sys::Syslog qw(:DEFAULT);
+use JSON::XS qw(decode_json);
```
Comment (floatingatoll) [inline Bugzilla.pm]: Is BMO generally fine with modules that require compilation? It seems reasonable to me, but in case that's a topic y'all have past decisions on.

File: Bugzilla.pm [hunk 3]
```diff
@@ -767,6 +768,23 @@ sub elastic {
     $class->process_cache->{elastic} //= Bugzilla::Elastic->new();
 }
 
+sub check_rate_limit {
+    my ($class, $name, $id) = @_;
+    my $params = Bugzilla->params;
+    if ($params->{rate_limit_active}) {
```
Comment (floatingatoll) [inline Bugzilla.pm]: Moving this specific test up one level to the function calling ```check_rate_limit``` would improve the performance of rate limiting checks, very very slightly.

Normally I wouldn't suggest micro-optimizations, but since this function is used in the anonymous user request paths, if there's heavy load, any improvement may be worth considering.

If you choose to pursue this, you'd need a comment indicating that the caller is responsible for that test.

File: Bugzilla.pm [hunk 4]
```diff
@@ -767,6 +768,23 @@ sub elastic {
...
+        my $rules = decode_json($params->{rate_limit_rules});
```
Comment (floatingatoll) [inline Bugzilla.pm]: Does an acceptable user experience occur if ```{rate_limit_rules}``` is invalid JSON?

File: Bugzilla.pm [hunk 5]
```diff
@@ -767,6 +768,23 @@ sub elastic {
...
+        my $limit = $rules->{$name};
+        unless ($limit) {
+             warn "no rules for $name!";
```
Comment (floatingatoll) [inline Bugzilla.pm]: Can the rules be tested for this property at startup time, rather than logging once per request?

File: Bugzilla.pm [hunk 6]
```diff
@@ -767,6 +768,23 @@ sub elastic {
...
+             return 0;
+        }
+        if (Bugzilla->memcached->should_rate_limit("$name:$id", @$limit)) {
```
Comment (floatingatoll) [inline Bugzilla.pm]: Does an acceptable user experience occur if $limit is not an array reference?

File: extensions/Needinfo/template/en/default/bug/needinfo.html.tmpl [hunk 7]
```diff
@@ -92,11 +92,13 @@ $(function() {
       $('#needinfo_from_container').removeClass('bz_default_hidden');
       $('#needinfo_from').prop('disabled', false);
       $('#needinfo_role_identity').text('');
+      $('needinfo_from_textfield').prop('required', true);
```
Comment (floatingatoll) [inline extensions/Needinfo/template/en/default/bug/needinfo.html.tmpl]: Are these changes meant to be included in this pull request? (They might go away if another one is merged, perhaps?)

File: t/css.t [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (floatingatoll) [inline t/css.t]: Is this file part of rate limiting, or another open / in-flight pull request?

File: template/en/default/admin/params/admin.html.tmpl [hunk 9]
```diff
@@ -23,6 +23,14 @@
    desc = "Set up account policies"
 %]
 
+[% rate_limit_rules_desc = BLOCK %]
+This parameter is a json object that allows for overriding the default rate limit.
+Each key takes an array in the form of [MAX_RATE, SECONDS] where MAX_RATE is the maximum
+number of requests that can occur over SECONDS. The default is [75, 60] or 75 requests
+over 60 seconds. Valid keys are <code>get_b[%''%]ug</code> which covers JSONRPC, XMLRPC, REST and BZAPI single
```
Comment (floatingatoll) [inline template/en/default/admin/params/admin.html.tmpl]: As an outside, I'd love to understand more about what these ```[%''%]``` blocks are defending against.

File: template/en/default/admin/params/admin.html.tmpl [hunk 10]
```diff
@@ -23,6 +23,14 @@
    desc = "Set up account policies"
 %]
 
+[% rate_limit_rules_desc = BLOCK %]
+This parameter is a json object that allows for overriding the default rate limit.
+Each key takes an array in the form of [MAX_RATE, SECONDS] where MAX_RATE is the maximum
```
Comment (floatingatoll) [inline template/en/default/admin/params/admin.html.tmpl]: Consider some rewording here for flow:

"This parameter is a json object. It must have one or more valid keys, whose values are each an array [M,S]. There are two valid keys: <code>get; <code>show."

File: template/en/default/admin/params/admin.html.tmpl [hunk 11]
```diff
@@ -23,6 +23,14 @@
...
+number of requests that can occur over SECONDS. The default is [75, 60] or 75 requests
```
Comment (floatingatoll) [inline template/en/default/admin/params/admin.html.tmpl]: The syntax description above does not offer any way for the user to select the default.

File: template/en/default/global/user-error.html.tmpl [hunk 12]
```diff
@@ -1812,6 +1812,10 @@
     The token you submitted does not exist, has expired, or has
     been canceled.
 
+  [% ELSIF error == "rate_limit" %]
+    [% title = "Rate Limit Exceeded" %]
+    You have exceeded the rate limit.
```
Comment (floatingatoll) [inline template/en/default/global/user-error.html.tmpl]: We may wish to consider further wording here. This is a one-time opportunity to communicate to a distracted coder that we need them to behave more politely towards our instance. We might offer a reasonable default rate limit, request they set the user-agent to their email address, and encourage them to try again after fifteen minutes. (It's okay to open this as a separate issue, as bonus wording shouldn't block shipping a rate limit.)

File: Bugzilla.pm (hunk 2 above)
Comment (dylanwh) [inline Bugzilla.pm]: that's already a dependency.

Comment (floatingatoll) [inline Bugzilla.pm]: Oh, sigh, you're right. I'm sorry.

File: template/en/default/admin/params/admin.html.tmpl (hunk 9 above)
Comment (dylanwh) [inline template/en/default/admin/params/admin.html.tmpl]: A stupid sanity check that prevents us from talking about "bugs"

File: Bugzilla.pm (hunk 4 above)
Comment (dylanwh) [inline Bugzilla.pm]: It isn't allowed to be invalid (no way of saving an invalid value) although maybe covering that is a good idea.

File: Bugzilla.pm (hunk 6 above)
Comment (dylanwh) [inline Bugzilla.pm]: Also checked by the config layer.

File: Bugzilla.pm (hunk 5 above)
Comment (dylanwh) [inline Bugzilla.pm]: It can change at any time -- we could cache it, but that would require clearing when the params are changed -- possibly memcached get_config() could work here. I'll talk to @umohm1  about that.

Comment (dylanwh) [inline Bugzilla.pm]: At the current time, there is no support for caching param values. So we'll deal with the performance for now.

Comment (dylanwh) [inline Bugzilla.pm]: See: https://bugzilla.mozilla.org/show_bug.cgi?id=1355463

File: template/en/default/admin/params/admin.html.tmpl (hunk 10 above)
Comment (umohm1) [inline template/en/default/admin/params/admin.html.tmpl]: I'll make the change.

File: template/en/default/admin/params/admin.html.tmpl (hunk 11 above)
Comment (umohm1) [inline template/en/default/admin/params/admin.html.tmpl]: Could you clarify a bit on this ?

File: Bugzilla/Config/Admin.pm [hunk 13]
```diff
@@ -44,11 +47,37 @@ sub get_param_list {
...
+sub check_rate_limit_rules {
+    my $rules = shift;
+
+    my $val = eval { decode_json($rules) };
+    return "failed to parse json" unless defined $val;
+    return "value is not HASH"    unless ref $val && ref($val) eq 'HASH';
+    return "rules are invalid"    unless all {
+        ref($_) eq 'ARRAY' && looks_like_number( $_->[0] ) && looks_like_number( $_->[1] )
+    } values %$val;
+
```
Comment (dylanwh) [inline Bugzilla/Config/Admin.pm]: This needs to return "missing get_bug" when get_bug is missing, and "missing show_bug" when show_bug is missing.

File: Bugzilla.pm [hunk 14]
```diff
@@ -767,6 +768,23 @@ sub elastic {
...
+            Bugzilla->audit("[rate_limit] $id exceeds rate limit $name: " . join("/", @$limit));
```
Comment (dylanwh) [inline Bugzilla.pm]: Just to cover here -- in the previous review glob wanted to add the remote_ip here, to protect against the future when $id might not be the remote ip address. We're specifically *not* doing this now because it will just make the log noisier. I'll either add it when we need it, or when we have structured logging. So this part is fine.

File: Bugzilla/Config/Admin.pm (hunk 13 above)
Comment (umohm1) [inline Bugzilla/Config/Admin.pm]: While I'm not sure what this code is doing. Would I be leaving this as is and adding in a conditional?

Comment (dylanwh) [inline Bugzilla/Config/Admin.pm]:
```perl
foreach my $required (qw( show_bug get_bug )) {
    return "missing $required" unless exists $val->{$required};
}
```

File: Bugzilla.pm (hunk 3 above)
Comment (dylanwh) [inline Bugzilla.pm]: Just a note that this performance thing was noted and at the moment ignored.
